src/org/oidc/service/base/ServiceContext.java

```diff
@@ -0,0 +1,328 @@
+     * At this time we know of 3 such modes:
+     * fragment, queryString, formPost
+     * In callback, we can keep the redirect uris per response mode
+     * separate.
+     */
+    private Map<DataLocation,String> callBack;
+
+    /**
+     * Constants
+     */
+    private static final String SIG = "sig";
+    private static final String RSA = "rsa";
+    private static final String SHA_256 = "SHA-256";
+    private static final String ISSUER = "issuer";
+
+    public ServiceContext(
+            KeyJar keyJar,
+            ServiceContextConfig config) {
+        this.keyJar = keyJar;
+        this.config = config;
+    }
+
+    public ServiceContext() {
+
+    }
+
+    /**
+     The client needs its own set of keys. It can either dynamically create them or load them from local storage. This method can also fetch other entities keys provided that the URL points to a JWKS.
+     * @param keySpecifications contains fileName and algorithm
+     **/
+        public void importKeys(Map<FileOrUrl,KeySpecifications> keySpecifications) {
+        if(keySpecifications == null) {
+            throw new IllegalArgumentException("null keySpecifications");
+        }
+
+        Set<FileOrUrl> keys = keySpecifications.keySet();
+        KeySpecifications keySpecificationsIndex;
+        Key rsaKey;
+        KeyBundle keyBundle;
+        for(FileOrUrl key : keys) {
+            if(FileOrUrl.FILE.equals(key)) {
+                keySpecificationsIndex = keySpecifications.get(key);
+                if(RSA.equalsIgnoreCase(keySpecificationsIndex.getAlgorithm())) {
+                    rsaKey = new RSAKey(Jwk.importPrivateRsaKeyFromFile(keySpecificationsIndex.getFileName()), SIG);
+                    keyBundle = new KeyBundle();
+                    keyBundle.addKey(rsaKey);
+                    keyJar.addKeyBundle("", keyBundle);
+                }
+            } else if (FileOrUrl.URL.equals(key)) {
+                keyBundle = new KeyBundle();
+                keyJar.addKeyBundle(ISSUER,keyBundle);
+            }
+        }
+    }
+
+    /**
+     A 1<->1 map is maintained between a URL pointing to a file and
+     the name of the file in the file system.
+
+     As an example if the base_url is 'https://example.com' and a jwks_uri
+     is 'https://example.com/jwks_uri.json' then the filename of the
+     corresponding file on the local filesystem would be 'jwks_uri'.
+     Relative to the directory from which the RP instance is run.
+     * @param webName the published URL
+     * @return local filename
+     **/
+    public String fileNameFromWebname(String webName) throws ValueException {
+        if(Strings.isNullOrEmpty(webName)) {
+            throw new IllegalArgumentException("null or empty webName");
+        }
+
+        if(!webName.startsWith(this.baseUrl)) {
+            throw new ValueException("Webname does not match baseUrl");
+        }
+
+        webName = webName.substring(this.baseUrl.length());
+        if(webName.startsWith("/")) {
```

Comment:
Using "/" doesn't seem alright. File paths are OS specific. I would propose that you use https://docs.oracle.com/javase/8/docs/api/java/io/File.html#separator

@rohe opinions?